x-pack/plugins/session_view/public/components/tty_player/hooks.ts

```diff
@@ -0,0 +1,230 @@
+ */
+export const useIOLines = (pages: ProcessEventsPage[] | undefined) => {
+  const lines: IOLine[] = useMemo(() => {
+    const newLines: IOLine[] = [];
+
+    if (!pages) {
+      return newLines;
+    }
+
+    return pages.reduce((previous, current) => {
+      if (current.events) {
+        current.events.forEach((event) => {
+          if (event?.process?.io?.text) {
+            const data: IOLine[] = event.process.io.text.split(/\n\r?/).map((line) => {
+              return {
+                value: line,
+              };
+            });
+
+            previous = previous.concat(data);
+          }
+        });
+      }
+
+      return previous;
+    }, newLines);
+  }, [pages]);
+
+  return lines;
+};
+
+export interface XtermPlayerDeps {
+  ref: React.RefObject<HTMLElement>;
+  isPlaying: boolean;
+  lines: IOLine[];
+  hasNextPage?: boolean;
+  fetchNextPage?: () => void;
+  isFullscreen?: boolean;
+}
+
+export const useXtermPlayer = ({
+  ref,
+  isPlaying,
+  lines,
+  hasNextPage,
+  fetchNextPage,
+  isFullscreen,
+}: XtermPlayerDeps) => {
+  const { euiTheme } = useEuiTheme();
+  const { font, colors } = euiTheme;
+  const [currentLine, setCurrentLine] = useState(0);
+  const [userSeeked, setUserSeeked] = useState(false);
+  const [playSpeed] = useState(DEFAULT_TTY_PLAYSPEED_MS); // potentially configurable
+
+  const [terminal, fitAddon, searchAddon] = useMemo(() => {
+    const term = new Terminal({
+      theme: {
+        background: 'rgba(0,0,0,0)',
+        selection: colors.warning,
+      },
+      fontFamily: font.familyCode,
+      fontSize: 11,
+      allowTransparency: true,
+    });
+
+    const fitInstance = new FitAddon();
+    const searchInstance = new SearchAddon();
+
+    term.loadAddon(fitInstance);
+    term.loadAddon(searchInstance);
+
+    return [term, fitInstance, searchInstance];
+  }, [colors, font]);
+
+  useEffect(() => {
+    if (ref.current) {
+      terminal.open(ref.current);
+    }
+  }, [terminal, ref]);
+
+  useEffect(() => {
+    // isFullscreen check is there just to avoid the necessary "unnecessary" react-hook dep
+    // When isFullscreen changes, e.g goes from false to true and vice versa, we need to call fit.
+    if (isFullscreen !== undefined) {
+      fitAddon.fit();
+    }
+  }, [isFullscreen, fitAddon]);
+
+  const render = useCallback(
+    (lineNumber: number) => {
+      if (lines.length === 0) {
+        return;
+      }
+
+      let linesToPrint;
+
+      if (userSeeked) {
+        linesToPrint = lines.slice(0, lineNumber);
+        terminal.clear();
+        setUserSeeked(false);
+      } else {
+        linesToPrint = [lines[lineNumber]];
+      }
+
+      linesToPrint.forEach((line, index) => {
+        if (line?.value !== undefined) {
+          terminal.writeln(line.value);
+        }
+      });
+    },
+    [terminal, lines, userSeeked]
+  );
+
+  useEffect(() => {
+    if (isPlaying) {
+      const timer = setTimeout(() => {
```

Comment:
Interesting, I was wondering why `setTimeout` over `setInterval`, but looks like updating `currentLine` and having it on the `deps` array it's making call itself. Side-effects playing hard here 😆

Reply:
Ya not sure if there is a more performant way, lots depends on that currentLine being updated, so seems like the way to go. It performs well in my tests. 
We can play with the DEFAULT_TTY_PLAYSPEED_MS const and/or the number of lines to increment by every setTimeout if it becomes an issue on slow machines.

Reply:
Good idea. Some perf test would be nice! We can add a TODO to run a profiler test later on, by setting`4x or 6x CPU slowdown` on throttling settings and take notes on performance degradation